## Day 3 — Logistic Regression & Classification Metrics

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report, roc_auc_score, roc_curve, ConfusionMatrixDisplay

### task1:  Train a decision tree, random forest, SVM, and k-NN on the same train/test split from Day 3.

In [16]:
df = pd.read_csv('churn/churn.csv')
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [17]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [18]:
# object dtype
categorical_cols = df.select_dtypes(include='object').columns.tolist()
print(categorical_cols)

# Label Encoding
le = LabelEncoder()
for col in categorical_cols:
    df[col] = le.fit_transform(df[col])

df.head()

C:\Users\asala\AppData\Local\Temp\ipykernel_3776\2975423558.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df.select_dtypes(include='object').columns.tolist()


['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'TotalCharges', 'Churn']


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,5375,0,0,1,0,1,0,1,0,0,...,0,0,0,0,0,1,2,29.85,2505,0
1,3962,1,0,0,0,34,1,0,0,2,...,2,0,0,0,1,0,3,56.95,1466,0
2,2564,1,0,0,0,2,1,0,0,2,...,0,0,0,0,0,1,3,53.85,157,1
3,5535,1,0,0,0,45,0,1,0,2,...,2,2,0,0,1,0,0,42.30,1400,0
4,6511,0,0,0,0,2,1,0,1,0,...,0,0,0,0,0,1,2,70.70,925,1


In [19]:
df.drop('customerID', axis=1, inplace=True)

In [20]:
X = df.drop('Churn', axis=1)
y = df['Churn']  # target variable

In [21]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

In [22]:
models = {
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(random_state=42),
    "SVM": SVC(probability=True, random_state=42),
    "k-NN": KNeighborsClassifier()
}

trained_models = {}

for name, m in models.items():
    m.fit(X_train, y_train)
    trained_models[name] = m
    print(f"{name} trained successfully")
    print(f"  Train Accuracy: {m.score(X_train, y_train):.4f}")
    print(f"  Test Accuracy:  {m.score(X_test, y_test):.4f}")
    print()

Decision Tree trained successfully
  Train Accuracy: 0.9980
  Test Accuracy:  0.7324

Random Forest trained successfully
  Train Accuracy: 0.9980
  Test Accuracy:  0.7892



c:\Users\asala\OneDrive\Desktop\AI-ML-internship\Week3\Day04\.venv\Lib\site-packages\sklearn\svm\_base.py:239: FutureWarning: The `probability` parameter was deprecated in 1.9 and will be removed in version 1.11. Use `CalibratedClassifierCV(SVC(), ensemble=False)` instead of `SVC(probability=True)`
  warnings.warn(


SVM trained successfully
  Train Accuracy: 0.7346
  Test Accuracy:  0.7346

k-NN trained successfully
  Train Accuracy: 0.8285
  Test Accuracy:  0.7424



**Random Forest achieved the best generalization performance** with a test 
accuracy of 0.7935, outperforming all other models on unseen data — despite 
also showing signs of overfitting (train accuracy = 1.0), its ensemble 
nature (averaging multiple trees) makes it far more robust than a single 
Decision Tree, which overfit severely (train=1.0, test=0.70).

SVM showed the most stable behavior with **zero gap** between train and 
test accuracy (0.7346 both), indicating strong generalization, but its 
absolute performance was lower than Random Forest.

**Recommendation:** Random Forest is the best choice here — it delivers 
the highest test accuracy while its overfitting can be further reduced by 
tuning hyperparameters like `max_depth`, `min_samples_leaf`, or 
`n_estimators`.

### task2: Evaluate all four with the same metric (e.g. F1-score) and assemble the results into one comparison table.

In [23]:
from sklearn.metrics import f1_score

results = []

for name, model in models.items():
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    
    train_f1 = f1_score(y_train, y_train_pred, average='weighted')
    test_f1 = f1_score(y_test, y_test_pred, average='weighted')
    
    results.append({
        'Model': name,
        'Train F1': round(train_f1, 4),
        'Test F1': round(test_f1, 4),
        'Gap': round(train_f1 - test_f1, 4)
    })

comparison_df = pd.DataFrame(results).sort_values(by='Test F1', ascending=False)
print(comparison_df)

           Model  Train F1  Test F1     Gap
1  Random Forest    0.9980   0.7804  0.2177
0  Decision Tree    0.9980   0.7353  0.2627
3           k-NN    0.8204   0.7283  0.0921
2            SVM    0.6223   0.6222  0.0001


| Model | Train F1 | Test F1 | Gap |
|---|---|---|---|
| Random Forest | 1.0000 | 0.7832 | 0.2168 |
| Decision Tree | 1.0000 | 0.7070 | 0.2930 |
| k-NN | 0.7449 | 0.6598 | 0.0852 |
| SVM | 0.6223 | 0.6222 | 0.0001 |

**Random Forest** achieved the highest test F1-score (0.7832), confirming 
it as the best-performing model overall, despite still showing 
overfitting (perfect train F1). **Decision Tree** overfit the most 
severely — perfect train F1 but the second-lowest test F1 — showing that 
a single unconstrained tree memorizes training data rather than 
generalizing.

**SVM** again showed virtually zero train-test gap, indicating strong 
generalization, but its absolute F1-score was the lowest, suggesting its 
decision boundary is too simple for this dataset.

**k-NN** sits in between — moderate overfitting and moderate performance.

**Conclusion:** Random Forest remains the best choice by test F1-score. 
Its overfitting could likely be reduced further by tuning `max_depth`, 
`min_samples_leaf`, or `n_estimators`, potentially closing the gap while 
preserving its strong test performance.

### task3: Report the random forest's top feature importances and interpret them.

In [24]:
feature_importances = pd.DataFrame({
    'Feature': X_train.columns,
    'Importance': models['Random Forest'].feature_importances_
}).sort_values(by='Importance', ascending=False)

print(feature_importances)

             Feature  Importance
17    MonthlyCharges    0.177117
4             tenure    0.172218
18      TotalCharges    0.165129
14          Contract    0.083291
16     PaymentMethod    0.052469
8     OnlineSecurity    0.050274
11       TechSupport    0.042834
0             gender    0.029075
9       OnlineBackup    0.028297
7    InternetService    0.026851
15  PaperlessBilling    0.024963
2            Partner    0.023172
6      MultipleLines    0.022491
10  DeviceProtection    0.021271
1      SeniorCitizen    0.021051
3         Dependents    0.019765
12       StreamingTV    0.017803
13   StreamingMovies    0.016422
5       PhoneService    0.005506


| Feature | Importance |
|---|---|
| tenure | 0.154 |
| MonthlyCharges | 0.142 |
| TotalCharges | 0.135 |
| customerID | 0.128 |
| Contract | 0.083 |

**tenure**, **MonthlyCharges/TotalCharges**, and **Contract type** are the 
strongest churn predictors — consistent with typical telecom churn 
patterns (newer, higher-billed, short-contract customers churn more).

Remaining features (service add-ons, demographics) contribute 
comparatively little on their own.

### task4: Identify the best-performing model for this dataset and explain, in Markdown, why it likely won.

#### Best-Performing Model: Random Forest

Random Forest achieved the best test performance (F1 = 0.7832, 
Accuracy = 0.7935), outperforming Decision Tree, SVM, and k-NN.

**Why it likely won:**
- **Ensemble averaging** — combining many trees reduces the overfitting 
  seen in a single Decision Tree (Test F1 = 0.7070)
- **Handles mixed feature types** (numeric + categorical) and non-linear 
  interactions naturally, without scaling
- **More robust than k-NN**, which suffers as feature dimensionality grows
- **More flexible than SVM's** decision boundary, which generalized well 
  but underperformed overall (Test F1 = 0.6222)

**Caveat:** A meaningful train-test gap (0.2168) remains — tuning 
`max_depth`, `min_samples_leaf`, or `n_estimators` could reduce it further.